### ISSUE TITLE
Build state on Followers

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#3894]
**Description**

I would like to bring this topic up again, because I'm not sure why we stopped talking about this and would like to challenge some ideas.

### Idea

It should be straight forward, just start the same actors but the processor for example uses a NoopWriter etc.

### Challenges

When we build state on followers we have the following challenges:

 * Slow Follower
 * Fast Follower
 * Exporters

#### Slow Follower

If a follower is slower then the leader with processing and becomes later leader, this is not really a problem. It has indeed a snapshot with lower process position, but this is the same as I would increase the snapshot interval. Our current reprocessing mechanism should handle that.

If the follower is to slow to receive events, then normally atomix sends snapshots from the leader.
Then we need to halt our actors. Install the snapshot and recover from it.

#### Fast Follower

We might think it is a problem that if the follower is faster to process then the leader, because it stands on higher position and no event was created for that. This is also not really a problem, because we will not take a snapshot or make it valid until the last written event position is reached.
This means if this fast follower becomes leader only valid snapshots are used on recovery, this position have been reached by the old leader as well otherwise the snapshot can't be valid.

##### Exporters

Exporters are the bigger challenge where we have to think how we want to handle that. There exist several options to do that.

  1. Run the exporters also on followers. This would be possible, since the exporters should be idempotent. But this will increase the load of the background systems and in general in the cluster.
  2. Build an endpoint, which the follower can query. This endpoint will just return the last exporter position. This can then be used by the follower to determine where it can delete.
  3. Publish the last exported position via gossip by the leader. Periodically we could send the last exported position for each partition. This information can then be consumed by the follower to delete data.

### Gains

We could solve our current snapshot replication topic and everything related to that. 
We could reduce code depth, since we do not need anything related to that anymore. Atomix will also handle snapshot replication anyway.  The speed of fail over would be probably similar to our current snapshot replication approach.

This will for example the following bug issues:
 
  * https://github.com/zeebe-io/zeebe/issues/3887 
  * https://github.com/zeebe-io/zeebe/issues/3889
  * https://github.com/zeebe-io/zeebe/issues/3763

### Drawbacks 

Depends on the exporter solution.

**Comment by npepinpe** [comment]
Thanks for the detailed write-up!

I think we'll have to build this at some point, as I don't see any reasonable solution other than this when dealing with large state - replicating states of tens of GB will require an increase in snapshot interval, which means a bigger log, which means slower failover, which means...etc.

I think the drawbacks to building state on the follower should be explored (e.g. performance impact? coordinating exporters?), but we should at least invest some time to do so.

I think for the exporters the gossip idea (or did you mean CRDT?) is a good first approach - it's easy to implement, cheap, and we already work under a "at-least-once-guarantee", so duplicates are expected. The downside with this approach is that on cluster-wide restart, how do you recover the position? It's still always safe to export from the recovery point, but I guess we'd want to optimize this to avoid doing too much unnecessary work. That said, this could be done as a follow up.

I think the other approaches work but their drawbacks are more prominent.

**Comment by Zelldon** [comment]
Thanks @npepinpe 

> I think for the exporters the gossip idea (or did you mean CRDT?) is a good first approach - it's easy to implement, cheap, and we already work under a "at-least-once-guarantee", so duplicates are expected. The downside with this approach is that on cluster-wide restart, how do you recover the position? It's still always safe to export from the recovery point, but I guess we'd want to optimize this to avoid doing too much unnecessary work. That said, this could be done as a follow up.

Yes I thought about gossip. I think recovery should work when the leaders directly send the last exporter position they have in there state?

Would be very happy If we can try this out. :slightly_smiling_face:

**Comment by Zelldon** [comment]
In raft disst. is is also described that it makes more sense to do the snapshots or build the state on followers as well.


> We considered an alternative leader-based approach in which only the leader would create a snapshot, then it would send this snapshot to each of its followers. However, this has two disadvantages. First, sending the snapshot to each follower would waste network bandwidth and slow the snapshotting process. Each follower already has the information needed to produce its own snapshots, and it is typically much cheaper for a server to produce a snapshot from its local state than it is to send and receive one over the network. Second, the leader's implementation would be more complex. For example, the leader would need to send snapshots to followers in parallel with replicating new log entries to them, so as not to block new client requests.

https://raft.github.io/raft.pdf page 12f

**Comment by Zelldon** [comment]
superseded by https://github.com/zeebe-io/enhancements/pull/2
